lib/remote-storage.js

```diff
@@ -99,16 +100,24 @@ module.exports = class RemoteStorage {
    */
   async uploadDir (dir, prefix, postFileUploadCallback) {
     if (typeof prefix !== 'string') throw new Error('prefix must be a valid string')
-    async function _filterFiles (files) {
-      const bools = await Promise.all(files.map(async f => (await fs.stat(f)).isFile()))
-      return files.filter(f => bools.shift())
-    }
 
-    const files = await _filterFiles((await fs.readdir(dir)).map(f => path.join(dir, f)))
+    // walk the whole directory recursively using klaw.
+    const files = []
+    for await (const file of klaw(dir)) {
+      if (file.stats.isFile()) {
+        files.push(file.path)
+      }
+    }
 
     // parallel upload
     return Promise.all(files.map(async f => {
-      const s3Res = await this.uploadFile(f, prefix)
+      // get file's relative folder to the base directory.
+      var prefixDirectory = path.dirname(path.relative(dir, f))
+      // base directory returns ".", ignore that.
+      prefixDirectory = prefixDirectory === '.' ? '' : prefixDirectory
+      // newPrefix is now the initial prefix plus the files relative directory path.
+      const newPrefix = utils.urlJoin(prefix, prefixDirectory)
+      const s3Res = await this.uploadFile(f, newPrefix)
       if (postFileUploadCallback) postFileUploadCallback(f)
```

Comment:
just a minor comment instead of `postFileUploadCallback(f)` I would do `postFileUploadCallback(path.relative(dir, f))`

Reply:
That would not do the effect you are looking for because the invocation of `uploadDir` looks like this:
https://github.com/adobe/aio-app-scripts/blob/master/scripts/deploy.ui.js#L54

so that needs to also change, doing that :)

Reply:
indeed good catch :)

Reply:
fixed in my latest commit

Reply:
A test is failing now
   Expected: "index.html"
   Received: "../../index.html"
This test should pass for both windows and linux

also `path.relative(dist, f)` should be `path.relative(this.config.web.src)`

Thanks !

Reply:
I fixed the test, the callback expects a full path
Also path.relative(dist, f) should NOT be  path.relative(this.config.web.src, f)
Because `uploadDir` is called on `dist` now they might be the same, but we do not care, we care that the same EXACT directory passed to `uploadDir` is also passed to `path.relative` since `f` in the callback is guaranteed to be a child of whatever directory is passed to `uploadDir`

Reply:
yup you are right ok all good, sorry was a bit quick on that one.
thanks for the contribution and bearing with the required changes!!
Merging now